include/irr/asset/material_compiler/IR.h

```diff
@@ -0,0 +1,429 @@
+            UTextureOrConstant<type_of_const> value;
+        };
+
+        _IRR_STATIC_INLINE_CONSTEXPR size_t MAX_CHILDREN = 16ull;
+        struct children_array_t {
+            INode* array[MAX_CHILDREN] {};
+            size_t count = 0ull;
+
+            inline bool operator!=(const children_array_t& rhs) const
+            {
+                if (count != rhs.count)
+                    return true;
+
+                for (uint32_t i = 0u; i < count; ++i)
+                    if (array[i]!=rhs.array[i])
+                        return true;
+                return false;
+            }
+            inline bool operator==(const children_array_t& rhs) const
+            {
+                return !operator!=(rhs);
+            }
+
+            inline bool find(E_SYMBOL s, size_t* ix = nullptr) const 
+            { 
+                auto found = std::find_if(begin(),end(),[s](const INode* child){return child->symbol==s;});
+                if (found != (array+count))
+                {
+                    if (ix)
+                        ix[0] = found-array;
+                    return true;
+                }
+                return false;
+            }
+
+            operator bool() const { return count!=0ull; }
+
+            INode** begin() { return array; }
+            const INode*const* begin() const { return array; }
+            INode** end() { return array+count; }
+            const INode*const* end() const { return array+count; }
+
+            inline INode*& operator[](size_t i) { assert(i<count); return array[i]; }
+            inline const INode* const& operator[](size_t i) const { assert(i<count); return array[i]; }
+        };
+        template <typename ...Contents>
+        static inline children_array_t createChildrenArray(Contents... children) 
+        { 
+            children_array_t a;
+            const INode* ch[]{ children... };
+            memcpy(a.array, ch, sizeof(ch));
+            a.count = sizeof...(children);
+            return a; 
+        }
+
+        using color_t = core::vector3df_SIMD;
+
+        explicit INode(E_SYMBOL s) : symbol(s) {}
+        virtual ~INode() = default;
+
+        children_array_t children;
+        E_SYMBOL symbol;
+        bool deinited = false;
+    };
+
+    struct CMaterialNode : public INode
+    {
+        CMaterialNode() : INode(ES_MATERIAL), thin(false)
+        {
+            opacity.source = EPS_CONSTANT;
+            opacity.value.constant = color_t(1.f);
+        }
+
+        SParameter<color_t> opacity;
+        bool thin;
+    };
+
+    struct CGeomModifierNode : public INode
+    {
+        enum E_TYPE
+        {
+            ET_DISPLACEMENT,
+            ET_HEIGHT,
+            ET_NORMAL,
+            ET_DERIVATIVE
+        };
+        enum E_SOURCE
+        {
+            ESRC_UV_FUNCTION,
+            ESRC_TEXTURE
+        };
+
+        CGeomModifierNode(E_TYPE t) : INode(ES_GEOM_MODIFIER), type(t) {}
+
+        E_TYPE type;
+        E_SOURCE source;
```

Comment:
then comment out the `source` if its still a TODO for Nvidia MDL

Reply:
? i dont get this comment

Reply:
dont have an unused variable handing around